### DISCUSSION TITLE
Insertion performance in gremlin mode

### DISCUSSION THREAD

**Discussion by tzagit** [ArcadeData/arcadedb#889] [Q&A]
Hello,

since this is my first message here, I just want to say that I'm evaluating ArcadeDB since a few days and so far it seems very promising.

I'm currently trying to migrate an application from Amazon Neptune database to another graph database.

The application is based on NodeJS and use Apache Tinkerpop Gremlin for querying the graph database.

We are migrating because while Neptune is a very good production ready graph database, well it is starting to cost too much money.

The other graph databases that I've tested:

- Datastax Enterprise: works well, seems production ready, nice tools, it was very simple to migrate. The problem is the cost of the license which is very high.
- JanusGraph: looked promising,  it was very simple to migrate, but in practice, according to my humble experience, it may be difficult to operate it in production. Simple things like create rebuild indexes where difficult for instance.
- NebulaGraph: technically it doesn't support Gremlin. I tried they're own proprietary language, it was OK for simple traversal, but not quite for more complex ones that I already have working in Gremlin. It seems to be pretty good performance wise though.
- ArcadeDB: looks promising, simple to deploy, multimodel offers different options to query the data which is awesome, nice tools, migration of my code is relatively simple, and so far I like it very much.

But, but, but...

I have a few problems making it work as well as it should.

So I need some help.

Since ArcadeDB doesn't support the nodejs gremlin driver, I'm using the HTTP REST API to send gremlin commands.

I need to load my database with a base graph, consisting simply consisting of two vertice types, let's say A and B, linked by an edge.

I have approximately 30.000 vertices of type A, and approximately 10.000.000 vertices of type B.

I also defined unique indexes, and full text indexes for types A and B.

The problem I have is when my load script start to send gremlin commands to the HTTP REST API, a lot of them fail with the "Concurrent modification of page" exception.

I retry the command until it pass, and it finally pass, but the problem is that the rate at which I can insert data in the graph is not very high.

It's like 30-40 vertices by second, but I have millions of them...

I think that may be this "Concurrent modification of page" exception is slowing down my load times.

Can you explain to me how to get rid of it and reach a good level of performance?

Thank you very much.

**Comment by lvca** [comment]
If you're using the HTTP API, are you waiting for the response to send the next command? In this way, the latency would kill performance, for sure. If this is a one-time import, why don't you use the console to execute the gremlin commands?


Example:

```
bin/console.sh -b "create database graph;create vertex type User;create property User.id long; create index on User(id) unique;set language = gremlin;load ./myfile.gremlin" 
```

**Comment by tzagit** [comment]
Hi,

here is what my script looks like:

- For one vertex of type A:
  - Insert or update Vertex A, wait for the response
  - then for a collection of vertices B:
    - insert one vertex B if it doesn't exist, wait for the response
    - add the edge between A and B if it doesn't exist, wait for the response
    - and iterate through the collection.

But of course since each hierarchy of A->B is independant of the others I try to parallelize the load for multiple collections of {A,{B}}.

It is working well with the other databases that I tried.

With ArcadeDB it seems that if I have two independant vertices that belongs to the same Bucket, then I reach the condition that at one moment ArcadeDB try to write multiple vertices data for the same Bucket and I get the concurrent exception error. At least it seems like it, I'm not sure how ArcadeDB works under the hood.

I tried to not parallelize the queries, but then the throughput is very low and loading my data may take too much time.

Also the script to load the data, is also the script that refreshes the data, so it must run continuously. The data are webscrapped from the internet, based on a list of rss feeds. Whenever the feed is updated, I need to update the database.

Of course after the initial load, the refresh takes less time since only a fraction of the rss feeds are updated between two refresh. With AWS Neptune a full refresh takes more or less one hour.

I can try to generate and load a gremlin script instead of sending REST commands, I will see if it improves things.

But my app is not limited to loading and refreshing data from the rss feeds.

The application is a social media app.

I have users that can:

- create an account
- log in
- fetch data from the database
- interact with the database in many different ways:
  - sending a post
  - sending a comment
  - liking a post
  - repost a post
  - etc.

All these actions will write data in the database and need to be done in parallel for client side performance reason. And given the way the writes are done by ArcadeDB, if I have multiple users connected at the same time I may encounter this concurrent modification exception if they end up in the same Bucket, I guess.

So the way I want to use ArcadeDB is like a realtime online database that can handle multiple users in parallel (thousands or more) with good response time.

GIven what I've read in the documentation it seems possible, so may be I am doing something wrong...

**Reply by lvca** [reply]
ArcadeDB, like many other DBMSs, doesn't use locks but rather an optimistic approach where if 2 concurrent transactions touch the same page, the latter gets a ConcurrentModificationException and has to retry the operation. Our Java driver hides this complexity by simply executing every command in a for loop of X retries. If ConcurrentModificationException is caught, simply recall the lambda with the user code. I forgot to mention that when a ConcurrentModificationException exception is thrown, the transaction is rolled back implicitly.

In ArcadeDB the concurrency is on pages (while with OrientDB was on single records). This dramatically improves performance with concurrent online processing.

So if you can wrap your calls to ArcadeDB (what language are you using) with a `for loop + try/catch` block, then the only issue is the batch processing, but you could upload a file with the commands and let the server execute them.

The issue with "send a request and wait for the response" is what kills performance, because even if ArcadeDB takes 0 ms to process the command, if the HTTP request takes 15ms to be sent and 15ms for the response is 30ms per request = 33 reqs/sec and there is nothing you can do with that, unless you send asynchronous requests or let the server to process the commands.

**Comment by tzagit** [comment]
Hi,

thank you for your answer.

Here is some additional information since  I've done other tests.

The hardware used for the server is an i7, 4 cores, 8 threads, with 8G of memory, the disk is a SATA SSD.

The ArcadeDB server runs in Docker.

I've profiled the queries to ArcadeDB, I launched multiple imports (on a sample of my rss feeds) and here are the results I've got:

My script is a nodejs script that run takes an arbitrary amount of rss feeds to parse and import in the database. The tasks are loaded into an array and I process 20 tasks in parallel from one nodejs process. It means that I have to open 20 connections to the database at peak.

For 20 concurrent tasks performing import in the database through HTTP REST API I have a ratio of failure of approximately 20 to 25% (measured on approximately 20000 requests attempts, not counting the retries). Whenever a query fails, if I got a concurrent modification exception I wait for one second then I retry the query, until it passes. In any other error cases I just drop the query.

So for the 20000 requests to succeed, I have to actually send 25000 (5000 will fail with a concurrent modification exception).

It means that in the vast majority of cases I finally succeed to write my data in the database, even if it may take several attempts.

Sometimes, however, I get strange security errors (user  is not allowed to write to bucket...). It's not a lot, but I'll investigate it later. I also got other  errors for duplicate key id, which is really strange since the keys I provide are unique (and in other databases I tried I never got this error).

In the meantime, I've adapted my scripts to work with OrientDB: the migration was very easy since it seems that the compatibility between ArcadeDB and OrientDB is very high. I just had to change the HTTP Rest url (and a minor gremlin fix, may be due to an older Tinkerpop version on OrientDB).

I've performed the same tests on OrientDB (in docker). With OrientDB, through HTTP REST API, I get absolutely no errors, but the CPU consumption jumps from 12% at max with ArcadeDB to 80% at max with OrientDB. The average CPU consumption is also way higher on OrientDB.

The throughput seems to be more or less the same though. I've got an error message on OrientDB when I tried to launch more than 20 threads to load the database, saying that I've reached the maximum number of connections in the gremlin pool in the server. I probably can tune that, but given that the CPU consumption is so high, I don't know if it's worth to test it.

So for 20 parallel tasks to load the database, it takes approximately the same time to perform the load on both databases. I got no error on OrientDB but the CPU consumption is way higher. The memory taken by the JVM was approximately the same in both databases.

I don't know what to say.

20 tasks to parallel load the base doesn't seem much. In a real case scenario with thousands of customers interacting with the database I wonder how it will perform. In one hand the CPU usage seems pretty low. On the other hand the multiple attempts to perform a write could cause long response time from the end user point of view.

Here are some metrics for the requests:

average duration: 235 ms
min duration: 5 ms
max duration: 6840 ms

Please note that this is the metrics for the succeeding http requests, it doesn't take into account the potentially multiple 1 second wait after failed attempts. So from the end user point of view, if it takes 2 attempts to succeed a write, the total time to wait would be 1,235 seconds on average.  In the current scenario, it would be the case for 20 to 25% of the requests. If I wait for a smaller time like 200 ms instead of one second, my ratio of failed query jump to 50% and my throughput is decreased.

In OrientDB in the same conditions I get these results:

 average duration: 397 ms
 min duration: 11 ms
 max duration: 6250 ms

So ArcadeDB has faster average response time and lower CPU consumption, but the necessary retries can add a lot of waiting time for the end user.

For the time taken for the HTTP roundtrips, do you plan to port the OrientDB nodejs driver to ArcadeDB? In the meantime I've adapted my scripts to take advantage of the postgres nodejs driver.

And the results of my tests are really strange.

Here they are:

in the same conditions, 20 parallel tasks to load the database I get absolutely no concurrent modifications exception, no duplicate keys exceptions, but I get a lot of error messages from the postgres nodejs connector.

How I use the connector: I create a pool of connections, when I send a command I get a client from the pool, send a begin transaction command, then send the command, in case of success I commit the command. Then I release the client to avoid connection leaks.

In case of failure, if the failure is related to the connection pool (connection timeout and transaction not began are the errors that I got the most), I release the client, I wait 200 ms and I retry the whole process (getting a client, opening a transaction, etc.).

The metrics are:

average duration: 141 ms
min duration: 8 ms
max duration: 8015 ms

The CPU and memory consumption look the same than with REST API.

So response time are better, but the throughput is lower, I didn't figure why yet. So in total it seems that it takes longer to load the database than with REST API, but without errors from the database. I need to re-check the throughput though.

A strange behavior that I have using the postgres connector: if I kill my load job, then ArcadeDB go crazy and takes 99% of CPU.

If I gracefully close the pool connector I don't have this problem. But in production, we can expect that some process client using the pool connector may crash and then cause the database to eat all the CPU. So this could be a problem.

So far I think that I have progressed using the postgres connector. Not having the thousands of concurrent exceptions is better.

But I need to understand better how to fine tune the postgres connector on client side, and may be on server side (with your help), to avoid the problem that I encountered: failures from the connection pool, and the way the database react when we kill a client process.

I hope that these infos are useful to you. Given the low CPU / memory consumption of ArcadeDB, if I can fix or fine tune the last problems, ArcadeDB can be a very promising database.

**Reply by lvca** [reply]
Thanks for the detailed response, very helpful. First of all, the ArcadeDB server runs with only 2GB with docker, by default. Did you change the default settings? Did you use 2GB of RAM with OrientDB too?

About the concurrent modification exception, you should avoid the `sleep` at all, all these operations take a few ms, so waiting for an entire second it's an eternity :-)

The Postgres driver could swallow some exceptions, which would explain why it's getting going with zero issues.

By the way, I fixed some issues on Postgres about 3 weeks ago, and one issue was about consuming the CPU in case of closing the connection, so that issue should be already fixed. Also, some issues about exception propagation were resolved around the same time.

What release are you using? Is there any chance you can use the last snapshot `23.2.1-SNAPSHOT` (build it from sources, it takes a few minutes, see https://github.com/ArcadeData/arcadedb/blob/main/CONTRIBUTING.md#developer-guide)? 

About the Node.js driver, we don't have plans to support it directly, but I'll be more than happy if some contributor can port the OrientDB one or write something simpler from scratch :-)

**Reply by lvca** [reply]
I forgot to mention, with high concurrency this setting on the server should improve response time: `-Darcadedb.profile=high-performance`

**Reply by lvca** [reply]
On docker, the default settings for memory are in the `ARCADEDB_OPTS_MEMORY` env variable:

From the Dockerfile:

```
ENV ARCADEDB_OPTS_MEMORY="-Xms2G -Xmx2G"
```

So you could just overwrite that and try 4GB or more (if you have enough physical RAM).
